File: tutorials/testing/python/bst.py
from bridges.bridges import *
from bridges.bst_element import *
from bridges.data_src_dependent.data_source import *
import sys

# recursive insert method to insert nodes into a binary search tree
def insertR(rt, newel):
    if (rt is None):
        return newel
    elif newel.get_key() < rt.get_key():
        rt.set_left(insertR(rt.get_left(), newel))
    else:
        rt.set_right(insertR(rt.get_right(), newel))
    return rt


def main():
    args = sys.argv[1:]

    # create the Bridges object, set credentials
    bridges = Bridges(int(args[0]), args[1], args[2])

    if len(args) > 3:
        bridges.connector.set_server(args[3])

    #Title and Description
    bridges.set_title("A Binary Search Tree Example with Earthquake Data")
    bridges.set_description("10 sets of earthquake are gathered and sorted with nodes whos magnitude is greater than "
        + "the parent is set to the right and nodes with a magnitude less than the parent set to the left. "
        + "The root is set to red the leaf nodes are neutral.")

    # Retrieve a list of 10 earthquake records from USGS using the BRIDGES API
    ami = get_earthquake_usgs_data(10)

    root = None
    # create BST nodes and insert into a tree
    for i in range(len(ami)):
        bst_node = BSTElement(ami[i].get_magnitude(), ami[i])
        # set label of the node
        bst_node.set_label(ami[i].get_title() + ami[i].get_time())

        root = insertR(root, bst_node)

    # set some visual attributes
    root.get_visualizer().set_color("red")

    #set visualizer type
    bridges.set_data_structure(root)
    #visualize the tree
    bridges.visualize()


if __name__ == "__main__":
    main()

File: tutorials/testing/c++/bst.cpp

#include <string>
#include "Bridges.h"
#include "DataSource.h"
#include "data_src/EarthquakeUSGS.h"
#include "BSTElement.h"

using namespace std;
using namespace bridges;
using namespace bridges::dataset;

BSTElement<float, EarthquakeUSGS> *insert (BSTElement<float, EarthquakeUSGS> *rt,
	BSTElement<float, EarthquakeUSGS> *new_el);

int max_quakes = 25;

int main(int argc, char **argv) {
	// create Bridges object
	Bridges bridges (YOUR_ASSSIGNMENT_NUMBER, "YOUR_USER_ID",
		"YOUR_API_KEY");

	// set title, description
	bridges.setTitle("A Binary Search Tree Example with Earthquake Data");
	bridges.setDescription("10 sets of earthquake are gathered and sorted with nodes whos magnitude is greater than the parent is set to the right and nodes with a magnitude less than the parent set to the left. The root is set to red the leaf nodes are neutral.");

	// get earthquake data
	DataSource *ds = new DataSource;
	vector<EarthquakeUSGS> eq_list = ds->getEarthquakeUSGSData(max_quakes);

	// create the binary tree root
	BSTElement<float, EarthquakeUSGS> *root = nullptr;

	// insert quake records into the tree
	for (int k = 0; k < max_quakes; k++) {
		EarthquakeUSGS eq = eq_list[k];
		BSTElement<float, EarthquakeUSGS>
		*bst_node = new BSTElement<float, EarthquakeUSGS>(eq.getMagnitude(), eq);
		bst_node->setLabel(eq.getTitle() + "\nLat/Long: ( " +
			to_string(eq.getLatit()) + "," + to_string(eq.getLongit()) + " )\n" +
			eq.getDateStr());
		root = insert (root, bst_node);
		root->setColor("red");
	}

	// visualize the binary search tree
	bridges.setDataStructure(root);
	bridges.visualize();

	return 0;
}

// inserts a record into the tree
BSTElement<float, EarthquakeUSGS> *insert (BSTElement<float, EarthquakeUSGS> *rt,
	BSTElement<float, EarthquakeUSGS> *new_el) {
	if (rt == nullptr)
		return new_el;
	else if (new_el->getKey() < rt->getKey())
		rt->setLeft(insert(rt->getLeft(), new_el));
	else
		rt->setRight(insert(rt->getRight(), new_el));

	return rt;
}


